compiler/src/vm/heap/object.rs

```diff
@@ -0,0 +1,238 @@
+use super::{pointer::Pointer, Heap};
+use crate::{
+    builtin_functions::BuiltinFunction,
+    compiler::{
+        hir_to_lir::HirToLir,
+        lir::{Instruction, Lir},
+    },
+    database::Database,
+    module::Module,
+};
+use itertools::Itertools;
+use num_bigint::BigInt;
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+#[derive(Clone)]
+pub struct Object {
+    pub reference_count: usize,
+    pub data: Data,
+}
+#[derive(Clone)]
+pub enum Data {
+    Int(Int),
+    Text(Text),
+    Symbol(Symbol),
+    Struct(Struct),
+    Closure(Closure),
+    Builtin(Builtin),
+}
+
+#[derive(Clone)]
+pub struct Int {
+    pub value: BigInt,
+}
+
+#[derive(Clone)]
+pub struct Text {
+    pub value: String,
+}
+
+#[derive(Clone)]
+pub struct Symbol {
+    // TODO: Choose a more efficient representation.
+    pub value: String,
+}
+
+#[derive(Default, Clone)]
+pub struct Struct {
+    // TODO: Choose a more efficient representation.
+    pub buckets: HashMap<u64, Vec<(Pointer, Pointer)>>, // hash -> list of keys and values
+}
+
+#[derive(Clone)]
+pub struct Closure {
+    pub captured: Vec<Pointer>,
+    pub num_args: usize,
+    pub body: Vec<Instruction>,
+}
+
+#[derive(Clone)]
+pub struct Builtin {
+    pub function: BuiltinFunction,
+}
+
+impl Struct {
+    pub fn from_fields(heap: &Heap, fields: HashMap<Pointer, Pointer>) -> Self {
+        let mut s = Self::default();
+        for (key, value) in fields {
+            s.insert(heap, key, value);
+        }
+        s
+    }
+    fn insert(&mut self, heap: &Heap, key: Pointer, value: Pointer) {
+        let bucket = self.buckets.entry(key.hash(heap)).or_default();
+        for (existing_key, existing_value) in bucket.iter_mut() {
+            if existing_key.equals(heap, key) {
+                *existing_value = value;
+                return;
+            }
+        }
+        bucket.push((key, value));
+    }
+    pub fn get(&self, heap: &Heap, key: Pointer) -> Option<Pointer> {
+        let bucket = self.buckets.get(&key.hash(heap))?;
+        for (existing_key, existing_value) in bucket {
+            if existing_key.equals(heap, key) {
+                return Some(*existing_value);
+            }
+        }
+        None
+    }
+    fn len(&self) -> usize {
+        self.buckets.values().map(|bucket| bucket.len()).sum()
+    }
+    pub fn iter(&self) -> impl Iterator<Item = (Pointer, Pointer)> {
+        self.buckets.clone().into_values().flatten()
+    }
+    fn equals(&self, heap: &Heap, other: &Struct) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        for (key_a, value_a) in self.iter() {
+            for (key_b, value_b) in other.iter() {
+                if key_a.equals(heap, key_b) && !value_a.equals(heap, value_b) {
```

Comment:
Different keys also mean that the structs are not equal

```suggestion
                if !key_a.equals(heap, key_b) || !value_a.equals(heap, value_b) {
```